### DISCUSSION TITLE
Example from Chapter 11 docs is incorrect?

### DISCUSSION THREAD

**Discussion by awa5114** [SCons/scons#4362] [Q&A] [answered]
I'm trying to implement one of the examples in Chapter 11 of the manual as follows:

```
|   hello.c
|   SConstruct.py
|   
\---install-dir
```

# `SConstruct.py`

```
env = Environment()
hello = env.Program('hello.c')
env.Install("#install-dir", hello)
env.Alias("install", "#install-dir")
```

# `hello.c`

```
int main()
{
    printf("Hello, world!\n");
}
```

According to the manual, running `scons -Q` should only build the executable and not install it. Only when I specify `scons -Q install` should the executable be installed in `install-dir`:

> This is an area where the `Alias` function comes in handy, allowing you, for example, to create a pseudo-target named `install` that can expand to the specified destination directory

However, in both cases where I use `scons -Q` and `scons -Q install` the executable gets installed. So I fail to see the purpose of `env.Alias()` in this example. What am I missing here?

**Comment by mwichmann** [comment, marked as answer]
This is all about target selection rules.  If `install` evaluates to something in the project tree (i.e. at or below the directory of the `SConstruct`, which it is in your case), *AND* you haven't given any targets on the cmdline or called the `Default` function, then that target will be one selected for building.  In the User Guide example, the directory to install to is `/usr/bin/`, which does not meet that selection rule, so it's not selected by default, only when you give the (alias) target on the cmdline.

**Reply by awa5114** [reply]
Thanks, you are correct, I changed the directory structure to:

```
+---dev
|       hello.c
|       SConstruct.py
|       
\---install-dir
```

and the `SConstruct.py` to 

```
env = Environment()
hello = env.Program('hello.c')
env.Install("../install-dir", hello)
env.Alias("install", "../install-dir")
```

And now the behavior is as described by the docs.

**Reply by awa5114** [reply]
@mwichmann I do have a follow up question actually. I'm now trying with the `InstallAs` function to change the name of the installed file as follows:

```
# Change the name of the file upon installation

env = Environment()
hello = env.Program("hello.c")
env.InstallAs("../install-dir/hello-new", hello)
env.Alias("install", "../install-dir")
```

When I build with `scons install` it does copy the executable file but doesn't add the `.exe` extension to it (I'm on Windows). I'm wondering why that is? Isn't SCons supposed to be platform independent? For the time being I've had to change it to the following for the executable to work:

```
# Change the name of the file upon installation

env = Environment()
hello = env.Program("hello.c")
env.InstallAs("../install-dir/hello-new.exe", hello)
env.Alias("install", "../install-dir")
```

**Reply by mwichmann** [reply]
With `InstallAs` you're providing the new name to use, the builder takes that very literally. If you want to have it be somewhat independent, you can use the construction variables `PROGSUFFIX` and `SHLIBSUFFIX`. Since your destination is a "target" it will have command substitution applied to it (you can't use these everywhere, only in an appropriate context):
```py
env.InstallAs("../install-dir/hello-new$PROGSUFFIX", hello)
```
See: https://scons.org/doc/production/HTML/scons-man.html#cv-PROGSUFFIX

To see what this is set to:
```py
print("DEBUG: progsuffix is", env['PROGSUFFIX'])
```
